sys/net/include/netdev.h

```diff
@@ -0,0 +1,262 @@
+     * @return  -ENOTSUP, if *opt* is not supported for the device with this
+     *          operation
+     * @return  -EOVERFLOW, if available space in *value* given in *value_len*
+     *          is not big enough to store the option value.
+     * @return  any other fitting negative errno if the ones stated above
+     *          are not sufficient
+     */
+    int (*get_option)(netdev_t *dev, netdev_opt_t opt, void *value,
+                      size_t *value_len);
+
+    /**
+     * @brief   Set an option value for a given network device.
+     *
+     * @param[in] dev           the network device
+     * @param[in] opt           the option type
+     * @param[in] value         the value to set
+     * @param[in] value_len     the length of *value*
+     *
+     * @return  0, on success
+     * @return  -EINVAL, if *value* is not in a required format
+     * @return  -ENODEV, if *dev* is not recognized
+     * @return  -ENOTSUP, if *opt* is not supported for the device with this
+     *          operation
+     * @return  -EOVERFLOW, if *value_len* is larger then the device expects.
+     * @return  -EPROTONOSUPPORT, if *opt* was NETDEV_OPT_PROTO and type is
+     *          not supported.
+     * @return  any other fitting negative errno if the ones stated above
+     *          are not sufficient
+     */
+    int (*set_option)(netdev_t *dev, netdev_opt_t opt, void *value,
+                      size_t value_len);
+
+    /**
+     * @brief   Get state from a given network device.
+     *
+     * @param[in] dev       the network device
+     * @param[out] state    the network device
+     *
+     * @return  0, on success
+     * @return  -ENODEV, if *dev* is not recognized
+     * @return  -ETIME, if device timed out on trying to acquire state
+     */
+    int (*get_state)(netdev_t *dev, netdev_state_t *state);
+
+    /**
+     * @brief   Set state from a given network device.
+     *
+     * @param[in] dev      the network device
+     * @param[in] action   action to trigger
+     *
+     * @return  0, on success
+     * @return  -ENODEV, if *dev* is not recognized
+     * @return  -ENOTSUP, if *state* is not supported
+     * @return  -ETIME, if device timed out on trying to change state
+     */
+    int (*trigger)(netdev_t *dev, netdev_action_t action);
+
+    /**
+     * @brief   Must be called by a controlling thread if a message of type
+     *          NETDEV_MSG_EVENT_TYPE was received
+     *
+     * @param[in] dev           the network device that fired the event.
+     * @param[in] event_type    Event type. Values are free to choose for the
+     *                          driver. Must be given in the @ref msg_t::value
+     *                          of the received message
+     */
+    void (*event)(netdev_t *dev, uint16_t event_type);
+} netdev_driver_t;
+
+struct netdev_t {
+    kernel_pid_t pid;           /**< the driver's thread's PID */
+    netdev_driver_t *driver;    /**< pointer to the devices interface */
```

Comment:
Each driver has a device descriptor (a struct containing the devices run-time data). This is the same concept used for many of our device driver (e.g. sensors, network interfaces, etc). The above `netdev_t` struct simply forces any network device descriptor to have these two parameters on the first positions. This way you can cast any device descriptor to act as `netdev_t` for being used by MAC layer implementations, while the device driver internally uses it's own type (e.g.`at86rf231_t`).